src/scripting/lua_kernel_base.cpp

```diff
@@ -446,6 +448,100 @@ static int intf_get_language(lua_State* L)
 	return 1;
 }
 
+static void dir_meta_helper(lua_State* L, std::vector<std::string>& keys)
+{
+	switch(luaL_getmetafield(L, -1, "__dir")) {
+		case LUA_TFUNCTION:
+			lua_pushvalue(L, 1);
+			lua_push(L, keys);
+			lua_call(L, 2, 1);
+			keys = lua_check<std::vector<std::string>>(L, -1);
+			break;
+		case LUA_TTABLE:
+			auto dir_keys = lua_check<std::vector<std::string>>(L, -1);
+			std::copy(dir_keys.begin(), dir_keys.end(), std::back_inserter(keys));
+			break;
+	}
+}
+
+/**
+ * Prints out a list of keys available in an object.
+ * A list of keys is gathered from the following sources:
+ * - For a table, all keys defined in the table
+ * - Any keys accessible through the metatable chain (if __index on the metatable is a table)
+ * - The output of the __dir metafunction
```

Comment:
How will this filter out deprecated stuff?

Replying to https://github.com/wesnoth/wesnoth/pull/5965#issuecomment-885810718 :
> Renaming something because the old name is misleading is an accomplishment in and of itself

Removing the old name is an accomplishment, but merely adding a new name gives people more things to look at. It will improve the readability of code that uses the new name, but its affect on the task of writing new code is not always beneficial.

Reply:
That's a good point. Perhaps I could add a `__dir_exclude` metafunction/table, which lists keys to exclude from the dir listing even if they exist in the table.

Reply:
I implemented something a bit different in the "Exclude deprecated things from dir()" commit (not direct-linking it because the hash could evolve as I force-push). Does that look sufficient to you? There could in theory be a few deprecated things that are caught by that system but they should be rare.

The implementation is likely inefficient, but as this is intended for use in the Lua console to discover API (particularly experimental API), I don't think that matters much.